src/lib/core/diag.c

```diff
@@ -158,17 +171,43 @@ error_format_msg(struct error *e, const char *format, ...)
 void
 error_append_msg(struct error *e, const char *format, ...)
 {
-	va_list ap;
-	va_start(ap, format);
-	int prefix_len = strlen(e->errmsg);
-	char *msg = e->errmsg + prefix_len;
-	vsnprintf(msg, sizeof(e->errmsg) - prefix_len, format, ap);
-	va_end(ap);
+	va_list ap1, ap2;
+	va_start(ap1, format);
+	va_copy(ap2, ap1);
+	int static_buf_size = sizeof(e->errmsg_buf);
+	int curr_len = strlen(e->errmsg);
+	int len = vsnprintf(NULL, 0, format, ap1);
+	int new_size = curr_len + len + 1;
+	va_end(ap1);
+
+	bool is_malloced = e->errmsg != e->errmsg_buf;
+	if (is_malloced || new_size > static_buf_size) {
+		char *new_buf = xmalloc(new_size);
+		memcpy(new_buf, e->errmsg, curr_len + 1);
+		if (is_malloced)
+			free(e->errmsg);
+		e->errmsg = new_buf;
+	}
+	char *msg = e->errmsg + curr_len;
+	int new_len = vsnprintf(msg, len + 1, format, ap2);
+	VERIFY(new_len == len);
+	va_end(ap2);
 }
 
 void
 error_vformat_msg(struct error *e, const char *format, va_list ap)
 {
-	vsnprintf(e->errmsg, sizeof(e->errmsg), format, ap);
+	/* Copy the `ap' to call `vsnprintf' twice. */
+	va_list ap_copy;
+	va_copy(ap_copy, ap);
+	error_msg_destroy(e);
+	int static_buf_size = sizeof(e->errmsg_buf);
+	int len = vsnprintf(e->errmsg, static_buf_size, format, ap);
+	bool is_truncated = len >= static_buf_size;
+	if (is_truncated) {
+		e->errmsg = xmalloc(len + 1);
+		int new_len = vsnprintf(e->errmsg, len + 1, format, ap_copy);
+		VERIFY(new_len == len);
```

Comment:
It does not look good that we have append and allocate buffer without exponential growth.

Also `error_msg_destroy` does not look good here. With amortized buffer we can try to reuse it.

Also as as Kostya mentioned the reason we have this buffer is not performance, may be we can reevaluate design and drop the preallocated buffer.

Reply:
> It does not look good that we have append and allocate buffer without exponential growth.

There is no sense in optimizing `error_append_msg()` for long messages. It is used in a few places, which likely will be switched to error payloads.

> Also error_msg_destroy does not look good here. With amortized buffer we can try to reuse it.

Removed `error_msg_destroy()`.

> Also as as Kostya mentioned the reason we have this buffer is not performance, may be we can reevaluate design and drop the preallocated buffer.

Let's keep the preallocated buffer.